tests/linearizability/model.go

```diff
@@ -244,25 +289,76 @@ func applyRequestToSingleState(s EtcdState, request EtcdRequest) (EtcdState, Etc
 	s.KeyValues = newKVs
 	opResp := make([]EtcdOperationResult, len(request.Ops))
 	increaseRevision := false
+
 	for i, op := range request.Ops {
 		switch op.Type {
 		case Get:
 			opResp[i].Value = s.KeyValues[op.Key]
 		case Put:
 			s.KeyValues[op.Key] = op.Value
 			increaseRevision = true
+			s = detachFromOldLease(s, op)
 		case Delete:
 			if _, ok := s.KeyValues[op.Key]; ok {
 				delete(s.KeyValues, op.Key)
 				increaseRevision = true
+				s = detachFromOldLease(s, op)
 				opResp[i].Deleted = 1
 			}
+		case PutWithLease:
+			if _, ok := s.Leases[op.LeaseID]; ok {
+				//handle put op.
+				s.KeyValues[op.Key] = op.Value
+				increaseRevision = true
+				s = detachFromOldLease(s, op)
+				s = attachToNewLease(s, op)
+			}
+		case LeaseRevoke:
+			//Delete the keys attached to the lease
+			keyDeleted := false
+			for key, _ := range s.Leases[op.LeaseID].Keys {
+				//same as delete.
+				if _, ok := s.KeyValues[key]; ok {
+					if !keyDeleted {
+						keyDeleted = true
+					}
+					delete(s.KeyValues, key)
+					delete(s.KeyLeases, key)
+				}
+			}
+			//delete the lease
+			delete(s.Leases, op.LeaseID)
+			if keyDeleted {
+				increaseRevision = true
+			}
+		case LeaseGrant:
+			lease := EtcdLease{
+				LeaseID: op.LeaseID,
+				Keys:    map[string]struct{}{},
+			}
+			s.Leases[op.LeaseID] = lease
 		default:
 			panic("unsupported operation")
 		}
 	}
+
 	if increaseRevision {
 		s.Revision += 1
 	}
+
 	return s, EtcdResponse{Result: opResp, Revision: s.Revision}
 }
+
+func detachFromOldLease(s EtcdState, op EtcdOperation) EtcdState {
+	if oldLeaseId, ok := s.KeyLeases[op.Key]; ok {
+		delete(s.Leases[oldLeaseId].Keys, op.Key)
+		delete(s.KeyLeases, op.Key)
+	}
+	return s
+}
+
+func attachToNewLease(s EtcdState, op EtcdOperation) EtcdState {
```

Comment:
```suggestion
func attachKeyToLease(s EtcdState, leaseID int64, key string) EtcdState {
```

Reply:
updated in https://github.com/etcd-io/etcd/pull/15093